### DISCUSSION TITLE
esp32 serial port

### DISCUSSION THREAD

**Discussion by Silaalis** [Moddable-OpenSource/moddable#1055] [Q&A]
code to use esp2 serial port 2 in Arduino is something like
```c
#define RXD2 13
 #define TXD2 12
 void setup() {
   Serial.begin(115200);
   Serial2.begin(115200, SERIAL_8N1, RXD2, TXD2);
   delay(1000);
   Serial.println("Loopback program started");
 }
 void loop() {
   if(Serial.available()){
     Serial.write("-");
     Serial2.write(Serial.read());  
   }
   if(Serial2.available()){
     Serial.write(".");
     Serial.write(Serial2.read());  
   }
 }
```
what is equivalent code in nodejs?

**Comment by phoddie** [comment]
> what is equivalent code in nodejs?

This repository is not about Node.js. Node.js does not run on an ESP32 or other embedded microcontrollers.

The Moddable SDK provides access to the serial port in JavaScript on an ESP32. It uses the [ECMA-419 I/O](https://419.ecma-international.org/#-9-io-class-pattern) APIs for that. There is a [serial echo example](https://github.com/Moddable-OpenSource/moddable/blob/public/examples/io/serial/echo/main.js) in the Moddable SDK that is similar to your example. Here's the relevant code snippet.

```js
new device.io.Serial({
	...device.Serial.default,
	baud: 115200,
	port: 2,
	format: "buffer",
	onReadable: function(count) {
		this.write(this.read());
	},
});
```
